storm-core/src/clj/backtype/storm/daemon/nimbus.clj

```diff
@@ -346,53 +400,104 @@
+                                                                     (atom min-replication-count))
+        current-replication-count-code (atom (get-blob-replication-count (master-stormcode-key storm-id) nimbus))
+        current-replication-count-conf (atom (get-blob-replication-count (master-stormconf-key storm-id) nimbus))
+        total-wait-time (atom 0)]
+    (log-message "wait for desired replication" "count"
+      min-replication-count "wait-time" max-replication-wait-time
+      "code" @current-replication-count-code
+      "conf" @current-replication-count-conf
+      "jar" @current-replication-count-jar
+      "replication count" (get-blob-replication-count (master-stormconf-key storm-id) nimbus))
+    (if (:blob-store nimbus)
+      (while (and (> min-replication-count @current-replication-count-jar)
+               (> min-replication-count @current-replication-count-code)
+               (> min-replication-count @current-replication-count-conf)
+               (or (= -1 max-replication-wait-time)
+                 (< @total-wait-time max-replication-wait-time)))
         (sleep-secs 1)
         (log-debug "waiting for desired replication to be achieved.
           min-replication-count = " min-replication-count  " max-replication-wait-time = " max-replication-wait-time
-          "current-replication-count = " @current-replication-count " total-wait-time " @total-wait-time)
+          (if (not (local-mode? conf))"current-replication-count for jar key = " @current-replication-count-jar)
+          "current-replication-count for code key = " @current-replication-count-code
+          "current-replication-count for conf key = " @current-replication-count-conf
+          " total-wait-time " @total-wait-time)
         (swap! total-wait-time inc)
-        (reset! current-replication-count  (.getReplicationCount (:code-distributor nimbus) storm-id))))
-  (if (< min-replication-count @current-replication-count)
-    (log-message "desired replication count "  min-replication-count " achieved,
-      current-replication-count" @current-replication-count)
-    (log-message "desired replication count of "  min-replication-count " not achieved but we have hit the max wait time "
-      max-replication-wait-time " so moving on with replication count = " @current-replication-count)
-    )))
-
-(defn- read-storm-topology [conf storm-id]
-  (let [stormroot (master-stormdist-root conf storm-id)]
-    (Utils/deserialize
-      (FileUtils/readFileToByteArray
-        (File. (master-stormcode-path stormroot))
-        ) StormTopology)))
+        (if (not (local-mode? conf))(reset! current-replication-count-conf  (get-blob-replication-count (master-stormconf-key storm-id))))
+          (reset! current-replication-count-code  (get-blob-replication-count (master-stormcode-key storm-id)))
+          (reset! current-replication-count-jar  (get-blob-replication-count (master-stormjar-key storm-id)))))
```

Comment:
I am finding this hard to read.  Could we indent the `if` and outdent the following `reset!` calls properly?